File: negocio/ListaDobleEnemigos.java
package negocio;

import modelo.Enemigo;

public class ListaDobleEnemigos {
    private NodoEnemigo primero;
    private NodoEnemigo ultimo;
    private int tamanio;

    public ListaDobleEnemigos() {
        this.primero = null;
        this.ultimo = null;
        this.tamanio = 0;
    }

    public void insertarAlFinal(Enemigo enemigo) {
        NodoEnemigo nuevo = new NodoEnemigo(enemigo);
        if (primero == null) {
            primero = nuevo;
            ultimo = nuevo;
        } else {
            ultimo.setSiguiente(nuevo);
            nuevo.setAnterior(ultimo);
            ultimo = nuevo;
        }
        tamanio++;
    }

    public boolean eliminarPorId(int id) {
        NodoEnemigo actual = primero;
        while (actual != null) {
            if (actual.getEnemigo().getId() == id) {
                if (actual == primero && actual == ultimo) {
                    primero = null;
                    ultimo = null;
                } else if (actual == primero) {
                    primero = primero.getSiguiente();
                    primero.setAnterior(null);
                } else if (actual == ultimo) {
                    ultimo = ultimo.getAnterior();
                    ultimo.setSiguiente(null);
                } else {
                    actual.getAnterior().setSiguiente(actual.getSiguiente());
                    actual.getSiguiente().setAnterior(actual.getAnterior());
                }
                tamanio--;
                return true;
            }
            actual = actual.getSiguiente();
        }
        return false;
    }

    public Enemigo buscarPorId(int id) {
        NodoEnemigo actual = primero;
        while (actual != null) {
            if (actual.getEnemigo().getId() == id) {
                return actual.getEnemigo();
            }
            actual = actual.getSiguiente();
        }
        return null;
    }

    public void recorrerAdelante() {
        if (primero == null) {
            System.out.println("No hay enemigos activos.");
            return;
        }
        NodoEnemigo actual = primero;
        while (actual != null) {
            System.out.println(actual.getEnemigo());
            actual = actual.getSiguiente();
        }
    }

    public void recorrerAtras() {
        if (ultimo == null) {
            System.out.println("No hay enemigos activos.");
            return;
        }
        NodoEnemigo actual = ultimo;
        while (actual != null) {
            System.out.println(actual.getEnemigo());
            actual = actual.getAnterior();
        }
    }

    public NodoEnemigo getPrimero() { return primero; }
    public int getTamanio() { return tamanio; }
}

File: Python/negocio/lista_doble_enemigos.py
from negocio.nodo_enemigo import NodoEnemigo

class ListaDobleEnemigos:
    """Lista doblemente enlazada manual para los enemigos activos en
    el campo, igual que en la versión Java."""

    def __init__(self):
        self.primero = None
        self.ultimo = None
        self.tamanio = 0

    def insertar_al_final(self, enemigo):
        nuevo = NodoEnemigo(enemigo)
        if self.primero is None:
            self.primero = nuevo
            self.ultimo = nuevo
        else:
            self.ultimo.siguiente = nuevo
            nuevo.anterior = self.ultimo
            self.ultimo = nuevo
        self.tamanio += 1

    def eliminar_por_id(self, id_):
        actual = self.primero
        while actual is not None:
            if actual.enemigo.id == id_:
                if actual is self.primero and actual is self.ultimo:
                    self.primero = None
                    self.ultimo = None
                elif actual is self.primero:
                    self.primero = self.primero.siguiente
                    self.primero.anterior = None
                elif actual is self.ultimo:
                    self.ultimo = self.ultimo.anterior
                    self.ultimo.siguiente = None
                else:
                    actual.anterior.siguiente = actual.siguiente
                    actual.siguiente.anterior = actual.anterior
                self.tamanio -= 1
                return True
            actual = actual.siguiente
        return False

    def buscar_por_id(self, id_):
        actual = self.primero
        while actual is not None:
            if actual.enemigo.id == id_:
                return actual.enemigo
            actual = actual.siguiente
        return None

    def recorrer_adelante(self):
        if self.primero is None:
            return ["No hay enemigos activos."]
        lineas = []
        actual = self.primero
        while actual is not None:
            lineas.append(str(actual.enemigo))
            actual = actual.siguiente
        return lineas

    def recorrer_atras(self):
        if self.ultimo is None:
            return ["No hay enemigos activos."]
        lineas = []
        actual = self.ultimo
        while actual is not None:
            lineas.append(str(actual.enemigo))
            actual = actual.anterior
        return lineas

    def get_primero(self):
        return self.primero

    def get_tamanio(self):
        return self.tamanio
